### PR TITLE
Implement fit_with() and formulas()

### PR DESCRIPTION
**Pull request by lionel-**
And start implementing formula-verbs: add_predictors()

### CODE REVIEW

**Comment by lionel-** [comment]
one test fails on Travis but not on my computer and I can't figure out why :/

File: R/formulas.R [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+fit_with <- function(data, .f, .formulas, ...) {
+  args <- list(...)
+
+  # Shadowing data to avoid getting the whole raw SEXP structure in
+  # the calls captured by the fitting function (which gets displayed
+  # in print methods)
+  args$data <- quote(data)
+
+  map(.formulas, function(formula) {
+    do.call(".f", c(list(formula), args))
```
Comment (hadley) [inline R/formulas.R]: This would be simpler with invoke, right? `invoke(.f, args, formula = formula)`

File: R/formulas.R [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  map(formulas, set_lhs, .response)
+}
+
+#' @rdname formulas
+#' @export
+formulae <- formulas
+
+validate_formulas <- function(response, formulas) {
+  if (!is_formula(response)) {
+    stop(".response should be a formula", call. = FALSE)
```
Comment (hadley) [inline R/formulas.R]: I think my convention is to use "must" instead of should

File: R/formulas.R [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  if (!length(formulas)) {
+    stop("No formula provided", call. = FALSE)
+  }
+
+  purrr::walk(formulas, function(f) {
+    if (!is_formula(f)) {
+      stop("'...' should only contain formulas", call. = FALSE)
```
Comment (hadley) [inline R/formulas.R]: I think "contain only" is better English, but I couldn't say why and I might be wrong

**Comment by hadley** [comment]
Overall, looks good. I suspect this API will evolve quite a bit over time, but this looks like a great start

File: R/formulas.R (hunk 1 above)
Comment (lionel-) [inline R/formulas.R]: yes but we'll need to adjust the environment used in `invoke()` then:

```r
invoke <- function(.f, .x = NULL, ...) {
  do.call(.f, c(.x, list(...)), envir = parent.frame())
}
```

How should we handle derived functions like `invoke_map()`? Probably no other way than adding `.env` as argument:

```r
invoke <- function(.f, .x = NULL, ..., .env = parent.frame()) {
  do.call(.f, c(.x, list(...)), envir = .env)
}
```

What do you think?

File: R/formulas.R (hunk 3 above)
Comment (lionel-) [inline R/formulas.R]: I may tell you about better French but not better English ;)

**Comment by lionel-** [comment]
One really good addition would be a `.save` argument taking a path to a .rds where we would save the dataset, the list of formulas and the fitted models. When that argument is supplied, we would only refit a model if the data or its formula changed.

Also we probably need to use `purrr::safely()` to fit the models.

That would be great for models that take hours to fit, especially now that we have rstanarm.

**Comment by lionel-** [comment]
Travis will fail until we merge the fix for `invoke()`

**Pull request merged**